File: libs/htmlcanvas/src/main/kotlin/mono/html/canvas/canvas/SelectionCanvasViewController.kt
package mono.html.canvas.canvas

import mono.graphics.geo.Rect
import mono.lifecycle.LifecycleOwner
import mono.livedata.LiveData
import mono.ui.theme.ThemeColor
import org.w3c.dom.HTMLCanvasElement
import org.w3c.dom.Path2D

/**
 * A canvas view controller to render the selection rectangle bound indicator.
 */
internal class SelectionCanvasViewController(
    lifecycleOwner: LifecycleOwner,
    canvas: HTMLCanvasElement,
    drawingInfoLiveData: LiveData<DrawingInfoController.DrawingInfo>
) : BaseCanvasViewController(canvas) {

    var selectingBound: Rect? = null

    init {
        drawingInfoLiveData.observe(lifecycleOwner, listener = ::setDrawingInfo)
    }

    override fun drawInternal() {
        val bound = selectingBound ?: return
        val leftPx = drawingInfo.toXPx(bound.left.toDouble())
        val topPx = drawingInfo.toYPx(bound.top.toDouble())
        val rightPx = drawingInfo.toXPx(bound.right + 1.0)
        val bottomPx = drawingInfo.toYPx(bound.bottom + 1.0)

        val path = Path2D().apply {
            moveTo(leftPx, topPx)
            lineTo(rightPx, topPx)
            lineTo(rightPx, bottomPx)
            lineTo(leftPx, bottomPx)
            closePath()
        }
        context.strokeStyle = ThemeColor.SelectionAreaStroke.colorCode
        context.lineWidth = 1.0
        context.setLineDash(DASH_PATTERN)
        context.stroke(path)
    }

    companion object {
        private val DASH_PATTERN = arrayOf(8.0, 6.0)
    }
}

File: monosketch-svelte/src/lib/mono/workspace/canvas/selection-canvas-view-controller.ts
import type { AppUiStateManager } from "$mono/ui-state-manager/app-ui-state-manager";
import { BaseCanvasViewController } from '$mono/workspace/canvas/base-canvas-controller';
import type { Rect } from '$libs/graphics-geo/rect';
import { ThemeColors } from '$mono/ui-state-manager/states';

const DASH_PATTERN = [8, 6];

/**
 * A canvas view controller to render the selection rectangle bound indicator.
 */
export class SelectionCanvasViewController extends BaseCanvasViewController {
    private selectingBound?: Rect;

    constructor(
        canvas: HTMLCanvasElement,
        private appUiStateManager: AppUiStateManager,
    ) {
        super(canvas);
    }

    setSelectingBound = (selectingBound?: Rect) => {
        this.selectingBound = selectingBound;
    };

    protected drawInternal() {
        const bound = this.selectingBound;
        if (!bound) {
            return;
        }

        const context = this.context;
        const drawingInfo = this.drawingInfo;

        const leftPx = drawingInfo.toXPx(bound.left);
        const topPx = drawingInfo.toYPx(bound.top);
        const rightPx = drawingInfo.toXPx(bound.right + 1.0);
        const bottomPx = drawingInfo.toYPx(bound.bottom + 1.0);

        const path = new Path2D();
        path.rect(leftPx, topPx, rightPx - leftPx, bottomPx - topPx);
        context.strokeStyle = this.appUiStateManager.getThemedColorCode(ThemeColors.SelectionAreaStroke);
        context.lineWidth = 1.0;
        context.setLineDash(DASH_PATTERN);
        context.stroke(path);
    }
}
